utils/linear_probes.py

```diff
@@ -0,0 +1,298 @@
+import os
+os.environ["OMP_NUM_THREADS"] = "6" # export OMP_NUM_THREADS
+os.environ["OPENBLAS_NUM_THREADS"] = "6" # export OPENBLAS_NUM_THREADS
+os.environ["MKL_NUM_THREADS"] = "6" # export MKL_NUM_THREADS
+os.environ["VECLIB_MAXIMUM_THREADS"] = "6" # export VECLIB_MAXIMUM_THREADS
+os.environ["NUMEXPR_NUM_THREADS"] = "6" # export NUMEXPR_NUM_THREADS
+
+from copy import copy
+from collections import defaultdict
+from typing import List, Dict, Tuple, Union, Optional
+
+import numpy as np
+import torch
+from transformer_lens import HookedTransformer, HookedTransformerConfig, FactoredMatrix, ActivationCache
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from einops import rearrange
+from sklearn.utils import shuffle
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import cross_val_score
+
+from data_generation.define_experiment import generate_variable_names, get_questions_dataset, randomly_swap_ents_to_vars
+# from src.lm_training_utils import linear_probe
+
+
+def get_activations(model: HookedTransformer, data: List[str]) -> Dict[str, List[np.ndarray]]:
+    acts_dict = defaultdict(list)
+    for d in data: # TODO consider batching
+        logits, activations = model.run_with_cache(d)
+        for act_str in activations.keys():
+            if 'hook_resid_post' in act_str: # only take activations after residual connection
+                acts_dict[act_str].append(copy(activations[act_str]).detach().cpu().numpy())
+    return acts_dict
+
+
+def train_linear_probe(x1, x2):
+    # concatenate the two datasets
+    x = rearrange([x1, x2], 'x n d -> (x n) d')
+    # labels: zero for data1, one for data2
+    y = rearrange([np.zeros(len(x1)), np.ones(len(x2))], 'x n -> (x n)')
+
+    x, y = shuffle(x, y, random_state=0)
+
+    # train a linear probe with l2 regularization and 5 fold cross validation
+    clf = LogisticRegression(random_state=0, max_iter=1000, penalty='l2', C=1.0)
+    scores = cross_val_score(clf, x, y, cv=5, scoring='accuracy', n_jobs=4)
+    return scores
+
+
+def leave_unique_q_type(data: List[str], model: HookedTransformer, q_type:str='born', filter_var_len=3) -> List[str]:
+    """
+    q_type is a string that must be contained in the question
+    model is needed for tokenization
```

Comment:
If model only needed for tokenization, it is better to provide tokenizer as argument to the function